lib/gringotts/gateways/sagepay.ex

```diff
@@ -0,0 +1,450 @@
+      commit_for_key(:post, "merchant-session-keys", vender_name, merchant_header)
+
+    merchant_key
+    |> Map.get("merchantSessionKey")
+  end
+
+  # card_params returns credit card details of a customer from Gringotts.Creditcard
+
+  defp card_params(card) do
+    %{
+      "cardDetails" => %{
+        "cardholderName" => CreditCard.full_name(card),
+        "cardNumber" => card.number,
+        "expiryDate" =>
+          (card.month * 100 + card.year)
+          |> Integer.to_string()
+          |> String.pad_leading(4, "0"),
+        "securityCode" => card.verification_code
+      }
+    }
+  end
+
+  # Function generate_card_identifier generate a unique cardidentifier for every transaction
+
+  defp generate_card_identifier(card, merchant_key) do
+    card_header = [
+      {"Authorization", "Bearer " <> merchant_key},
+      {"Content-type", "application/json"}
+    ]
+
+    card = card |> Poison.encode!()
+
+    {:ok, card_identifier} = commit_for_key(:post, "card-identifiers", card, card_header)
+
+    card_identifier
+    |> Map.get("cardIdentifier")
+  end
+
+  # Function transaction_details creates the actual body (details of the customer )of the card 
+  # and with merchant_session_key, card_identifiier ,shipping address of a
+  # customer, and other details and converting the map into keyword list
+
+  defp transaction_details(amount, merchant_key, card_identifiier, opts) do
+    {currency, value} = Money.to_string(amount)
+
+    Poison.encode!(%{
+      "transactionType" => opts[:transactionType],
+      "paymentMethod" => %{
+        "card" => %{
+          "merchantSessionKey" => merchant_key,
+          "cardIdentifier" => card_identifiier,
+          "save" => false
+        }
+      },
+      "vendorTxCode" => opts[:vendorTxCode],
+      "amount" => Kernel.trunc(String.to_float(value)),
+      "currency" => currency,
+      "description" => opts[:description],
+      "customerFirstName" => opts[:customerFirstName],
+      "customerLastName" => opts[:customerLastName],
+      "billingAddress" => %{
+        "address1" => opts[:billingAddress].address1,
+        "city" => opts[:billingAddress].city,
+        "postalCode" => opts[:billingAddress].postalCode,
+        "country" => opts[:billingAddress].country
+      }
+    })
+  end
+
+  # Parses sagepay's response and returns a `Gringotts.Response` struct
+  # in a `:ok`, `:error` tuple.
+
+  defp format_response(response) do
+    case response do
+      {:ok, %HTTPoison.Response{body: body}} -> {:ok, body |> Poison.decode!()}
+      _ -> %{"error" => "something went wrong, please try again later"}
+    end
+  end
+
+  @spec respond(term) :: {:ok | :error, Response}
+
+  defp respond({:ok, %{status_code: 201, body: body}}) do
+    body = body |> Poison.decode!()
+
+    {:ok,
+     %Response{
+       success: true,
+       id: body["transactionId"],
+       status_code: 201,
+       message: body["statusDetail"],
+       raw: body |> Poison.encode!()
```

Comment:
You were forced to do this _just_ because you didn't use a different var name on L426 :stuck_out_tongue_closed_eyes: